File: codex_token_trace/reports.py
from __future__ import annotations

import json
from collections import defaultdict
from typing import Any

from .formatting import fmt_int, fmt_pct, shorten, table
from .models import Interval, SessionAnalysis, ToolCall


def _event_type_totals(analysis: SessionAnalysis) -> tuple[dict[str, int], dict[str, int], dict[str, int]]:
    totals: dict[str, int] = defaultdict(int)
    counts: dict[str, int] = defaultdict(int)
    maxes: dict[str, int] = defaultdict(int)
    for event in analysis.events:
        key = event.type_key
        totals[key] += event.raw_chars
        counts[key] += 1
        maxes[key] = max(maxes[key], event.raw_chars)
    return totals, counts, maxes


def _top_interval_event_types(interval: Interval, limit: int = 3, width: int = 80) -> str:
    parts = []
    for key, chars, count in interval.top_event_types(limit=limit):
        parts.append(f"{key}:{fmt_int(chars)}({count})")
    return shorten(", ".join(parts), width)


def session_summary(analysis: SessionAnalysis) -> str:
    final = analysis.final_total
    compacted_chars = sum(event.raw_chars for event in analysis.events if event.top_type == "compacted")
    compacted_count = sum(1 for event in analysis.events if event.top_type == "compacted")
    rows = [
        ("session", shorten(analysis.session_id, 72)),
        ("path", str(analysis.path)),
        ("events", fmt_int(len(analysis.events))),
        ("token events", f"{fmt_int(len(analysis.unique_token_events))} unique / {fmt_int(len(analysis.token_events))} raw"),
        ("total tokens", fmt_int(final.total_tokens)),
        ("input tokens", fmt_int(final.input_tokens)),
        ("cached input tokens", fmt_int(final.cached_input_tokens)),
        ("output tokens", fmt_int(final.output_tokens)),
        ("reasoning output tokens", fmt_int(final.reasoning_output_tokens)),
        ("max input per event", fmt_int(analysis.max_input_tokens)),
        ("max non-cached input", fmt_int(analysis.max_non_cached_input_tokens)),
        ("max context usage", fmt_pct(analysis.max_context_ratio)),
        ("session_meta chars", f"{fmt_int(analysis.session_meta_chars_total)} total / {fmt_int(analysis.session_meta_chars_max)} max / {fmt_int(analysis.session_meta_count)} events"),
        ("tool output chars", fmt_int(analysis.tool_output_chars)),
        ("exec command output chars", fmt_int(analysis.exec_command_output_chars)),
        ("compacted chars", f"{fmt_int(compacted_chars)} total / {fmt_int(compacted_count)} events"),
    ]
    return table(("metric", "value"), rows)


def sessions_table(analyses: list[SessionAnalysis], limit: int | None = None) -> str:
    rows = []
    sorted_sessions = sorted(analyses, key=lambda a: a.final_total.total_tokens, reverse=True)
    if limit:
        sorted_sessions = sorted_sessions[:limit]
    for a in sorted_sessions:
        rows.append((
            shorten(a.session_id, 48),
            fmt_int(a.final_total.total_tokens),
            fmt_int(a.final_total.input_tokens),
            fmt_int(a.final_total.cached_input_tokens),
            fmt_int(a.final_total.output_tokens),
            fmt_int(a.max_non_cached_input_tokens),
            fmt_pct(a.max_context_ratio),
            fmt_int(a.tool_output_chars),
        ))
    return table(
        ("session", "total", "input", "cached", "output", "max_new_input", "max_ctx", "tool_out_chars"),
        rows,
    )


def tool_output_table(tools: list[ToolCall], limit: int = 30, *, only_exec: bool = False) -> str:
    rows = []
    filtered = [tool for tool in tools if tool.output_chars > 0]
    if only_exec:
        filtered = [tool for tool in filtered if tool.name == "exec_command"]
    for c in sorted(filtered, key=lambda c: c.output_chars, reverse=True)[:limit]:
        rows.append((
            fmt_int(c.output_chars),
            c.output_events,
            c.display_name,
            shorten(c.display_command, 100),
        ))
    return table(("output_chars", "events", "tool", "command_or_input"), rows)


def aggregate_tool_output_table(analyses: list[SessionAnalysis], limit: int = 30, *, only_exec: bool = False) -> str:
    by_key: dict[tuple[str, str], dict[str, int]] = defaultdict(lambda: {"output": 0, "count": 0})
    for analysis in analyses:
        for c in analysis.tools:
            if c.output_chars <= 0:
                continue
            if only_exec and c.name != "exec_command":
                continue
            key = (c.display_name, c.display_command)
            by_key[key]["output"] += c.output_chars
            by_key[key]["count"] += 1

    rows = []
    for (tool, command), values in sorted(by_key.items(), key=lambda kv: kv[1]["output"], reverse=True)[:limit]:
        rows.append((fmt_int(values["output"]), values["count"], tool, shorten(command, 100)))
    return table(("output_chars", "calls", "tool", "command_or_input"), rows)


def timeline_table(analysis: SessionAnalysis, limit: int | None = None) -> str:
    rows = []
    events = analysis.unique_token_events
    if limit:
        events = events[:limit]
    for e in events:
        rows.append((
            e.line_no,
            e.timestamp or "-",
            fmt_int(e.last.input_tokens),
            fmt_int(e.last.cached_input_tokens),
            fmt_int(e.last.non_cached_input_tokens),
            fmt_int(e.last.output_tokens),
            fmt_int(e.last.reasoning_output_tokens),
            fmt_int(e.last.total_tokens),
            fmt_pct(e.context_ratio),
            fmt_pct(e.last.cache_hit_rate),
        ))
    return table(
        ("line", "time", "input", "cached", "new_input", "output", "reasoning", "total", "ctx", "cache"),
        rows,
    )


def interval_table(analysis: SessionAnalysis, limit: int = 30) -> str:
    intervals = sorted(
        analysis.intervals,
        key=lambda interval: interval.token_event.last.non_cached_input_tokens,
        reverse=True,
    )[:limit]
    rows = []
    for interval in intervals:
        token = interval.token_event
        top_tool = interval.tools[0] if interval.tools else None
        rows.append((
            token.line_no,
            token.timestamp or "-",
            fmt_int(token.last.non_cached_input_tokens),
            fmt_int(token.last.input_tokens),
            fmt_int(token.last.cached_input_tokens),
            fmt_int(interval.tool_output_chars),
            fmt_int(interval.exec_command_output_chars),
            fmt_int(interval.event_count),
            _top_interval_event_types(interval, limit=3, width=88),
            shorten(top_tool.display_command if top_tool else "", 80),
        ))
    return table(
        (
            "line",
            "time",
            "new_input",
            "input",
            "cached",
            "prev_tool_out",
            "prev_exec_out",
            "prev_events",
            "top_prev_event_types",
            "top_prev_tool_or_command",
        ),
        rows,
    )


def drivers_table(analysis: SessionAnalysis) -> str:
    totals, counts, maxes = _event_type_totals(analysis)
    rows = []
    for key, chars in sorted(totals.items(), key=lambda kv: kv[1], reverse=True):
        rows.append((fmt_int(chars), counts[key], fmt_int(maxes[key]), key))
    return table(("raw_chars", "events", "max_chars", "event_type"), rows)


def compaction_table(analysis: SessionAnalysis, limit: int = 20) -> str:
    rows = []
    events = [event for event in analysis.events if event.top_type == "compacted" or event.payload_type == "context_compacted"]
    for event in sorted(events, key=lambda e: e.raw_chars, reverse=True)[:limit]:
        rows.append((event.line_no, event.timestamp or "-", fmt_int(event.raw_chars), event.type_key))
    return table(("line", "time", "raw_chars", "event_type"), rows)


def session_to_dict(analysis: SessionAnalysis) -> dict[str, Any]:
    totals, counts, maxes = _event_type_totals(analysis)
    return {
        "session_id": analysis.session_id,
        "path": str(analysis.path),
        "events": len(analysis.events),
        "token_events": len(analysis.unique_token_events),
        "raw_token_events": len(analysis.token_events),
        "final_total": analysis.final_total.__dict__,
        "max_input_tokens": analysis.max_input_tokens,
        "max_non_cached_input_tokens": analysis.max_non_cached_input_tokens,
        "max_context_ratio": analysis.max_context_ratio,
        "session_meta": {
            "chars_total": analysis.session_meta_chars_total,
            "chars_max": analysis.session_meta_chars_max,
            "count": analysis.session_meta_count,
        },
        "tool_output_chars": analysis.tool_output_chars,
        "exec_command_output_chars": analysis.exec_command_output_chars,
        "drivers": [
            {"event_type": key, "raw_chars": totals[key], "events": counts[key], "max_chars": maxes[key]}
            for key in sorted(totals, key=lambda k: totals[k], reverse=True)
        ],
        "heaviest_tools": [
            {
                "tool": tool.display_name,
                "command_or_input": tool.display_command,
                "output_chars": tool.output_chars,
                "output_events": tool.output_events,
                "timestamp": tool.timestamp,
            }
            for tool in sorted([t for t in analysis.tools if t.output_chars > 0], key=lambda t: t.output_chars, reverse=True)[:50]
        ],
        "intervals": [
            {
                "line": interval.token_event.line_no,
                "timestamp": interval.token_event.timestamp,
                "new_input": interval.token_event.last.non_cached_input_tokens,
                "input": interval.token_event.last.input_tokens,
                "cached": interval.token_event.last.cached_input_tokens,
                "tool_output_chars": interval.tool_output_chars,
                "exec_command_output_chars": interval.exec_command_output_chars,
                "event_count": interval.event_count,
                "top_event_types": [
                    {"event_type": key, "raw_chars": chars, "events": count}
                    for key, chars, count in interval.top_event_types(limit=5)
                ],
                "top_tool": interval.tools[0].display_command if interval.tools else None,
            }
            for interval in sorted(analysis.intervals, key=lambda i: i.token_event.last.non_cached_input_tokens, reverse=True)[:50]
        ],
    }


def analyses_to_json(analyses: list[SessionAnalysis], *, session: SessionAnalysis | None = None) -> str:
    payload: dict[str, Any]
    if session is not None:
        payload = {"session": session_to_dict(session)}
    else:
        payload = {"sessions": [session_to_dict(a) for a in analyses]}
    return json.dumps(payload, ensure_ascii=False, indent=2)

File: src/reports.ts
import { formatInt, formatPct, shorten, table } from "./formatting.js";
import {
  cacheHitRate,
  contextRatio,
  type EventRecord,
  eventTypeKey,
  execCommandOutputChars,
  finalTotal,
  type Interval,
  intervalExecCommandOutputChars,
  intervalToolOutputChars,
  isExecCommand,
  maxContextRatio,
  maxInputTokens,
  maxNonCachedInputTokens,
  nonCachedInputTokens,
  type SessionAnalysis,
  type TokenEvent,
  type ToolCall,
  type ToolUsage,
  toolDisplayName,
  toolInputLabel,
  toolOutputChars,
  topIntervalEventTypes,
} from "./models.js";

type EventTypeTotal = Readonly<{
  eventType: string;
  rawChars: number;
  events: number;
  maxChars: number;
}>;

type LargeEvent = Readonly<{
  line: number;
  timestamp: string | null;
  eventType: string;
  rawChars: number;
  preview: string;
}>;

type CompactionImpact = Readonly<{
  line: number;
  timestamp: string | null;
  rawChars: number;
  eventType: string;
  before: TokenEvent | null;
  after: TokenEvent | null;
}>;

export type ToolUsageSort = "output-chars" | "next-new-input";

export type JsonReportOptions = Readonly<{
  largeEventMinChars?: number;
  limit?: number;
}>;

const defaultJsonReportOptions = {
  largeEventMinChars: 10_000,
  limit: 50,
} as const;

const jsonReportOptions = (options: JsonReportOptions = {}): Required<JsonReportOptions> => ({
  largeEventMinChars: options.largeEventMinChars ?? defaultJsonReportOptions.largeEventMinChars,
  limit: options.limit ?? defaultJsonReportOptions.limit,
});

const eventTypeTotals = (analysis: SessionAnalysis): EventTypeTotal[] => {
  const totals = new Map<string, { rawChars: number; events: number; maxChars: number }>();
  for (const event of analysis.events) {
    const key = eventTypeKey(event);
    const current = totals.get(key) ?? { rawChars: 0, events: 0, maxChars: 0 };
    current.rawChars += event.rawChars;
    current.events += 1;
    current.maxChars = Math.max(current.maxChars, event.rawChars);
    totals.set(key, current);
  }
  return [...totals.entries()]
    .map(([eventType, value]) => ({ eventType, ...value }))
    .sort((a, b) => b.rawChars - a.rawChars);
};

const previewEvent = (event: EventRecord, width = 120): string => {
  const payload = event.raw.payload;
  if (typeof payload === "object" && payload !== null && !Array.isArray(payload)) {
    for (const key of ["message", "text", "output", "content", "input", "arguments"] as const) {
      const value = payload[key];
      if (typeof value === "string") {
        return shorten(value, width);
      }
    }
  }
  return shorten(JSON.stringify(event.raw), width);
};

export const largeEvents = (
  analysis: SessionAnalysis,
  minChars: number,
  limit: number,
): LargeEvent[] =>
  analysis.events
    .filter((event) => event.rawChars >= minChars)
    .map((event) => ({
      line: event.lineNo,
      timestamp: event.timestamp,
      eventType: eventTypeKey(event),
      rawChars: event.rawChars,
      preview: previewEvent(event),
    }))
    .sort((a, b) => b.rawChars - a.rawChars)
    .slice(0, limit);

const findPreviousToken = (analysis: SessionAnalysis, line: number): TokenEvent | null => {
  const previous = analysis.uniqueTokenEvents.filter((event) => event.lineNo < line).at(-1);
  return previous ?? null;
};

const findNextToken = (analysis: SessionAnalysis, line: number): TokenEvent | null => {
  const next = analysis.uniqueTokenEvents.find((event) => event.lineNo > line);
  return next ?? null;
};

export const compactionImpacts = (analysis: SessionAnalysis, limit: number): CompactionImpact[] =>
  analysis.events
    .filter((event) => event.topType === "compacted" || event.payloadType === "context_compacted")
    .map((event) => ({
      line: event.lineNo,
      timestamp: event.timestamp,
      rawChars: event.rawChars,
      eventType: eventTypeKey(event),
      before: findPreviousToken(analysis, event.lineNo),
      after: findNextToken(analysis, event.lineNo),
    }))
    .sort((a, b) => b.rawChars - a.rawChars)
    .slice(0, limit);

const usageBrief = (event: TokenEvent | null): string => {
  if (event === null) {
    return "-";
  }
  return `line ${event.lineNo}: input ${formatInt(event.last.inputTokens)}, cached ${formatInt(
    event.last.cachedInputTokens,
  )}, new ${formatInt(nonCachedInputTokens(event.last))}, total ${formatInt(event.total.totalTokens)}`;
};

export const sessionSummary = (analysis: SessionAnalysis): string => {
  const final = finalTotal(analysis);
  const compacted = compactionImpacts(analysis, Number.MAX_SAFE_INTEGER);
  const compactedChars = compacted.reduce((sum, event) => sum + event.rawChars, 0);
  const rows = [
    ["session", shorten(analysis.sessionId, 72)],
    ["path", analysis.path],
    ["events", formatInt(analysis.events.length)],
    [
      "token events",
      `${formatInt(analysis.uniqueTokenEvents.length)} unique / ${formatInt(analysis.tokenEvents.length)} raw`,
    ],
    ["total tokens", formatInt(final.totalTokens)],
    ["input tokens", formatInt(final.inputTokens)],
    ["cached input tokens", formatInt(final.cachedInputTokens)],
    ["output tokens", formatInt(final.outputTokens)],
    ["reasoning output tokens", formatInt(final.reasoningOutputTokens)],
    ["max input per event", formatInt(maxInputTokens(analysis))],
    ["max non-cached input", formatInt(maxNonCachedInputTokens(analysis))],
    ["max context usage", formatPct(maxContextRatio(analysis))],
    [
      "session_meta chars",
      `${formatInt(analysis.sessionMetaCharsTotal)} total / ${formatInt(
        analysis.sessionMetaCharsMax,
      )} max / ${formatInt(analysis.sessionMetaCount)} events`,
    ],
    ["tool output chars", formatInt(toolOutputChars(analysis))],
    ["exec command output chars", formatInt(execCommandOutputChars(analysis))],
    [
      "compacted chars",
      `${formatInt(compactedChars)} total / ${formatInt(compacted.length)} events`,
    ],
  ];
  return table(["metric", "value"], rows);
};

export const driversTable = (analysis: SessionAnalysis): string =>
  table(
    ["raw_chars", "events", "max_chars", "event_type"],
    eventTypeTotals(analysis).map((row) => [
      formatInt(row.rawChars),
      row.events,
      formatInt(row.maxChars),
      row.eventType,
    ]),
  );

export const largeEventsTable = (
  analysis: SessionAnalysis,
  minChars: number,
  limit: number,
): string =>
  table(
    ["line", "time", "raw_chars", "event_type", "preview"],
    largeEvents(analysis, minChars, limit).map((event) => [
      event.line,
      event.timestamp ?? "-",
      formatInt(event.rawChars),
      event.eventType,
      shorten(event.preview, 100),
    ]),
  );

export const compactionTable = (analysis: SessionAnalysis, limit: number): string =>
  table(
    ["line", "time", "raw_chars", "event_type", "before", "after"],
    compactionImpacts(analysis, limit).map((event) => [
      event.line,
      event.timestamp ?? "-",
      formatInt(event.rawChars),
      event.eventType,
      shorten(usageBrief(event.before), 88),
      shorten(usageBrief(event.after), 88),
    ]),
  );

const toolRows = (
  tools: ToolCall[],
  limit: number,
  onlyExec: boolean,
): Array<Array<string | number>> =>
  tools
    .filter((tool) => tool.outputChars > 0)
    .filter((tool) => !onlyExec || isExecCommand(tool))
    .sort((a, b) => b.outputChars - a.outputChars)
    .slice(0, limit)
    .map((tool) => [
      formatInt(tool.outputChars),
      tool.outputEvents,
      toolDisplayName(tool),
      shorten(toolInputLabel(tool), 100),
    ]);

export const toolOutputTable = (tools: ToolCall[], limit: number, onlyExec = false): string =>
  table(["output_chars", "events", "tool", "input_preview"], toolRows(tools, limit, onlyExec));

const sortedToolUsages = (toolUsages: ToolUsage[], sort: ToolUsageSort): ToolUsage[] =>
  [...toolUsages].sort((a, b) => {
    if (sort === "next-new-input") {
      return (b.nextNonCachedInputTokens ?? -1) - (a.nextNonCachedInputTokens ?? -1);
    }
    return b.outputChars - a.outputChars;
  });

const toolUsageRows = (
  toolUsages: ToolUsage[],
  limit: number,
  sort: ToolUsageSort,
): Array<Array<string | number>> =>
  sortedToolUsages(toolUsages, sort)
    .slice(0, limit)
    .map((usage) => [
      formatInt(usage.outputChars),
      toolDisplayName(usage),
      usage.startLine ?? usage.endLine ?? "-",
      usage.startTime ?? usage.endTime ?? "-",
      formatInt(usage.nextNonCachedInputTokens),
      formatPct(usage.nextNewInputRatio),
      formatInt(usage.nextInputTokens),
      formatInt(usage.nextCachedInputTokens),
      formatInt(usage.nextTotalTokens),
      shorten(toolInputLabel(usage), 100),
    ]);

export const toolUsageTable = (
  toolUsages: ToolUsage[],
  limit: number,
  sort: ToolUsageSort = "output-chars",
): string =>
  table(
    [
      "output_chars",
      "tool",
      "line",
      "time",
      "next_new_input",
      "next_new_ratio",
      "next_input",
      "next_cached",
      "next_total",
      "input_preview",
    ],
    toolUsageRows(toolUsages, limit, sort),
  );

export const aggregateToolOutputTable = (
  analyses: SessionAnalysis[],
  limit: number,
  onlyExec = false,
): string => {
  const byKey = new Map<
    string,
    { outputChars: number; calls: number; tool: string; input: string }
  >();
  for (const analysis of analyses) {
    for (const tool of analysis.tools) {
      if (tool.outputChars <= 0 || (onlyExec && !isExecCommand(tool))) {
        continue;
      }
      const key = `${toolDisplayName(tool)}\0${toolInputLabel(tool)}`;
      const current = byKey.get(key) ?? {
        outputChars: 0,
        calls: 0,
        tool: toolDisplayName(tool),
        input: toolInputLabel(tool),
      };
      current.outputChars += tool.outputChars;
      current.calls += 1;
      byKey.set(key, current);
    }
  }
  return table(
    ["output_chars", "calls", "tool", "input_preview"],
    [...byKey.values()]
      .sort((a, b) => b.outputChars - a.outputChars)
      .slice(0, limit)
      .map((row) => [formatInt(row.outputChars), row.calls, row.tool, shorten(row.input, 100)]),
  );
};

export const timelineTable = (analysis: SessionAnalysis, limit: number | null): string => {
  const events =
    limit === null ? analysis.uniqueTokenEvents : analysis.uniqueTokenEvents.slice(0, limit);
  return table(
    [
      "line",
      "time",
      "input",
      "cached",
      "new_input",
      "output",
      "reasoning",
      "total",
      "ctx",
      "cache",
    ],
    events.map((event) => [
      event.lineNo,
      event.timestamp ?? "-",
      formatInt(event.last.inputTokens),
      formatInt(event.last.cachedInputTokens),
      formatInt(nonCachedInputTokens(event.last)),
      formatInt(event.last.outputTokens),
      formatInt(event.last.reasoningOutputTokens),
      formatInt(event.last.totalTokens),
      formatPct(contextRatio(event)),
      formatPct(cacheHitRate(event.last)),
    ]),
  );
};

const topEventTypesText = (interval: Interval, limit: number, width: number): string => {
  const parts = topIntervalEventTypes(interval, limit).map(
    (event) => `${event.eventType}:${formatInt(event.rawChars)}(${event.events})`,
  );
  return shorten(parts.join(", "), width);
};

export const intervalTable = (analysis: SessionAnalysis, limit: number): string => {
  const intervals = [...analysis.intervals]
    .sort(
      (a, b) => nonCachedInputTokens(b.tokenEvent.last) - nonCachedInputTokens(a.tokenEvent.last),
    )
    .slice(0, limit);
  return table(
    [
      "line",
      "time",
      "new_input",
      "input",
      "cached",
      "prev_tool_out",
      "prev_exec_out",
      "prev_events",
      "top_prev_event_types",
      "top_prev_tool",
      "top_prev_input_preview",
    ],
    intervals.map((interval) => {
      const topTool = interval.tools[0] ?? null;
      return [
        interval.tokenEvent.lineNo,
        interval.tokenEvent.timestamp ?? "-",
        formatInt(nonCachedInputTokens(interval.tokenEvent.last)),
        formatInt(interval.tokenEvent.last.inputTokens),
        formatInt(interval.tokenEvent.last.cachedInputTokens),
        formatInt(intervalToolOutputChars(interval)),
        formatInt(intervalExecCommandOutputChars(interval)),
        formatInt(interval.eventCount),
        topEventTypesText(interval, 3, 88),
        topTool === null ? "" : toolDisplayName(topTool),
        topTool === null ? "" : shorten(toolInputLabel(topTool), 72),
      ];
    }),
  );
};

export const sessionsTable = (analyses: SessionAnalysis[], limit: number): string =>
  table(
    ["session", "total", "input", "cached", "output", "max_new_input", "max_ctx", "tool_out_chars"],
    [...analyses]
      .sort((a, b) => finalTotal(b).totalTokens - finalTotal(a).totalTokens)
      .slice(0, limit)
      .map((analysis) => [
        shorten(analysis.sessionId, 48),
        formatInt(finalTotal(analysis).totalTokens),
        formatInt(finalTotal(analysis).inputTokens),
        formatInt(finalTotal(analysis).cachedInputTokens),
        formatInt(finalTotal(analysis).outputTokens),
        formatInt(maxNonCachedInputTokens(analysis)),
        formatPct(maxContextRatio(analysis)),
        formatInt(toolOutputChars(analysis)),
      ]),
  );

const toolUsageToJson = (usage: ToolUsage): Record<string, unknown> => ({
  callId: usage.callId,
  tool: toolDisplayName(usage),
  inputPreview: toolInputLabel(usage),
  inputChars: usage.inputChars,
  outputChars: usage.outputChars,
  outputEvents: usage.outputEvents,
  startLine: usage.startLine,
  startTime: usage.startTime,
  endLine: usage.endLine,
  endTime: usage.endTime,
  nextTokenLine: usage.nextTokenLine,
  nextTokenTime: usage.nextTokenTime,
  nextInputTokens: usage.nextInputTokens,
  nextCachedInputTokens: usage.nextCachedInputTokens,
  nextNonCachedInputTokens: usage.nextNonCachedInputTokens,
  nextNewInputRatio: usage.nextNewInputRatio,
  nextOutputTokens: usage.nextOutputTokens,
  nextTotalTokens: usage.nextTotalTokens,
});

export const sessionToJson = (
  analysis: SessionAnalysis,
  options: JsonReportOptions = {},
): Record<string, unknown> => {
  const resolved = jsonReportOptions(options);
  return {
    sessionId: analysis.sessionId,
    path: analysis.path,
    events: analysis.events.length,
    tokenEvents: analysis.uniqueTokenEvents.length,
    rawTokenEvents: analysis.tokenEvents.length,
    finalTotal: finalTotal(analysis),
    maxInputTokens: maxInputTokens(analysis),
    maxNonCachedInputTokens: maxNonCachedInputTokens(analysis),
    maxContextRatio: maxContextRatio(analysis),
    sessionMeta: {
      charsTotal: analysis.sessionMetaCharsTotal,
      charsMax: analysis.sessionMetaCharsMax,
      count: analysis.sessionMetaCount,
    },
    toolOutputChars: toolOutputChars(analysis),
    execCommandOutputChars: execCommandOutputChars(analysis),
    drivers: eventTypeTotals(analysis),
    largeEvents: largeEvents(analysis, resolved.largeEventMinChars, resolved.limit),
    compactions: compactionImpacts(analysis, resolved.limit).map((event) => ({
      line: event.line,
      timestamp: event.timestamp,
      rawChars: event.rawChars,
      eventType: event.eventType,
      beforeLine: event.before?.lineNo ?? null,
      afterLine: event.after?.lineNo ?? null,
    })),
    toolUsages: analysis.toolUsages.slice(0, resolved.limit).map(toolUsageToJson),
    heaviestTools: analysis.tools
      .filter((tool) => tool.outputChars > 0)
      .sort((a, b) => b.outputChars - a.outputChars)
      .slice(0, resolved.limit)
      .map((tool) => ({
        tool: toolDisplayName(tool),
        inputPreview: toolInputLabel(tool),
        outputChars: tool.outputChars,
        outputEvents: tool.outputEvents,
        timestamp: tool.timestamp,
      })),
    intervals: [...analysis.intervals]
      .sort(
        (a, b) => nonCachedInputTokens(b.tokenEvent.last) - nonCachedInputTokens(a.tokenEvent.last),
      )
      .slice(0, resolved.limit)
      .map((interval) => ({
        line: interval.tokenEvent.lineNo,
        timestamp: interval.tokenEvent.timestamp,
        newInput: nonCachedInputTokens(interval.tokenEvent.last),
        input: interval.tokenEvent.last.inputTokens,
        cached: interval.tokenEvent.last.cachedInputTokens,
        toolOutputChars: intervalToolOutputChars(interval),
        execCommandOutputChars: intervalExecCommandOutputChars(interval),
        eventCount: interval.eventCount,
        topEventTypes: topIntervalEventTypes(interval, 5),
        topTool: interval.tools[0] ? toolDisplayName(interval.tools[0]) : null,
        topToolInputPreview: interval.tools[0] ? toolInputLabel(interval.tools[0]) : null,
      })),
  };
};

export const analysesToJson = (
  analyses: SessionAnalysis[],
  session?: SessionAnalysis,
  options: JsonReportOptions = {},
): string => {
  const payload = session
    ? { session: sessionToJson(session, options) }
    : { sessions: analyses.map((analysis) => sessionToJson(analysis, options)) };
  return JSON.stringify(payload, null, 2);
};
